Core/Epaper/Src/epaperUtils.c

```diff
@@ -788,131 +901,150 @@ void Paint_InitBmsImage(uint8_t* emptyImage)
 void Paint_DrawTableData(float data, DATA_TABLE_COL col, DATA_TABLE_ROW row)
 {
     // Set starting X and Y cordinates
-    uint16_t startX = DATA_START_X + (col * DATA_SEPERATION_X);
-    uint16_t startY = DATA_START_Y + (row * DATA_SEPERATION_Y);
-
-    // Clear data entry space
-    Paint_ClearWindows(startX, startY, startX + Font16.Width * 5, startY + Font16.Height, WHITE);
-
-    // Wrap to printable bounds
-    if(data > 9999.0f)
-    {
-        data = 9999.0f;
-    }
-    else if(data < -999.0f)
+    uint16_t startX = DATA_START_X;
+    for(int32_t i = 0; i < col; i++)
     {
-        data = -999.0f;
+        startX += tableSpacingCol[i];
     }
-
-    // Data position adjustment variables
-    uint8_t digits = 0;
-    bool negative = 0;
-
-    // Determine if data is negative and flip sign if neccesary
-    // Minus sign will take up one digit space
-    if(data < 0)
-    {
-        negative = true;
-        data *= -1;
-        digits++;
-    }
-
-    // Calucalte number of digits before decimal place
-    digits += calculateIntegerDigits(data);
-
-    bool decimal = false;
-    if(digits < 3)
-    {
-        decimal = true;
-        (col == DATA_VOLTAGE) ? (digits += 3) : (digits += 2);
-    }
-
-    startX += (Font16.Width / 2 ) * (4 - digits);
-
-    if(negative)
-    {
-        Paint_DrawString_EN(startX, startY, "-", &Font16, WHITE, BLACK);
-        startX += Font16.Width;
-    }
-
-    // Draw number as integer or decimal depending on number of digits
-    if(decimal)
-    {
-        if(col == DATA_VOLTAGE)
-        {
-            Paint_DrawNumDecimals(startX, startY, data, &Font16, 2, BLACK, WHITE);
-        }
-        else
-        {
-            Paint_DrawNumDecimals(startX, startY, data, &Font16, 1, BLACK, WHITE);
-        }
-    }
-    else
+    uint16_t startY = DATA_START_Y;
+    for(int32_t i = 0; i < row; i++)
     {
-        Paint_DrawNum(startX, startY, data, &Font16, BLACK, WHITE);
+        startY += tableSpacingRow[i];
     }
 
-    // Set data label to display
+    // Set label
+    uint8_t label;
+    uint32_t printWidth;
     if(col == DATA_VOLTAGE)
     {
-        Paint_DrawString_EN(startX + (digits * Font16.Width) - (negative ? Font16.Width : 0), startY, "V", &Font16, WHITE, BLACK);
+        label = 'V';
+        printWidth = 6;
     }
     else
     {
-        Paint_DrawString_EN(startX + (digits * Font16.Width) - (negative ? Font16.Width : 0), startY, "C", &Font16, WHITE, BLACK);
+        label = 'C';
+        printWidth = 5;
     }
-	
+
+    Paint_DrawCenteredLabeledFloat(data, label, printWidth, &Font16, startX, startY);
 }
 
 /*!
   @brief    Update BMS Image with current SOC
   @param    SOC BMS State of Charge as a percentage 
 */
-void Paint_DrawSOC(uint32_t SOC)
+void Paint_DrawSOE(uint32_t SOE)
 {
-	// Check bounds of SOC
-	if(SOC < 0)
+	// Check bounds of SOE
+	if(SOE < 0)
```

Comment:
If SOE != 100 then display 1 decimal digit. Ex 99.8% or 12.5%. See if we have room for the extra character